keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,656 @@
+1. `conditions`: the latest available observation of this `BucketAccess`. (Phase is not used.)
+1. `Type`: only one type is supported: "AccessGranted". If the condition's `Status` is true then the bucket is available to be accessed for the `Reason` provided. If `Status` is false then the bucket is unable to be accessed and, again, `Reason` states why.
+1. `Status`: one of true, false or unknown. If not specified then unknown is assumed. A true status indicates the this condition is true at the time of observation.
+1. `Reason`: one of "Available", "Bound", "Granting", "Revoking".  Note: "Available" indicates the `BucketAccess` was observed and its `conditions` was empty, which is expected when a brownfield `BucketAccess` is manually created.
+1. `Message`: optional human readable message containing more details.
+
+#### BucketAccessClass
+
+A cluster-scoped custom resource providing a way for admins to specify policies that may be used to access buckets.  A `BucketAccessClass` (BAC) is always applicable in greenfield, where access is dynamically granted, and only sometimes applicable in brownfield, where a user's identity may already exist, but not yet have access to a bucket.  In this case, a `BucketAccessRequest` will still specify the `BucketAccessClass` in order to determine which actions it is granted.
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: BucketAccessClass
+metadata: 
+  name:
+provisioner: [1]
+policyActions: [2]
+  allow:
+  - "*"
+  deny:
+  - "*"
+parameters: [3]
+```
+
+1. `provisioner`: (required) the name of the driver that `BucketAccess` instances should be managed by. Format: <provisioner-namespace>"/"<provisioner-name>, eg "ceph-rgw-provisoning/ceph-rgw.cosi.ceph.com".
+1. `policyActions`: a set of provisioner/platform defined policy actions to allow or deny a given user identity.
+1. `parameters`: (Optional)  A map of string:string key values.  Allows admins to control user and access provisioning by setting provisioner key-values.
+
+---
+
+### App Pod
+The application pod utilizes CSI's inline empheral volume support to provide the endpoint and secret credentials in an in-memory volume. This approach also, importantly, prevents the pod from launching before the bucket has been provisioned since the kubelet waits to start the pod until it has received the cosi-node-adpater's `NodePublishVolume` response.
+
+Here is a sample pod manifest:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: my-app-pod
+  namespace: dev-ns
+spec:
+  serviceAccountName: [1]
+  containers:
+    - name: my-app
+      image: ...
+      volumeMounts:
+        - mountPath: /cosi [2]
+          name: cosi-vol
+  volumes:
+    - name: cosi-vol
+      csi: [3]
+        driver: cosi.sigs.k8s.io [4]
+        volumeAttributes: [5]
+          bucketRequestName: <my-br-name>
+          bucketAccessRequestName: <my-bar-name>
+          bucketRequestNamespace:
```

Comment:
Does this mean a Pod can request/get access to a BR/BAR from another Namespace? That's a bit surprising.

Reply:
Good catch, this should not be here. @jeffvance --^

Reply:
II believe the cosi adapter needs this namespace. @wlan0 ? The key name could be more generic if it's confusing to see a BAR ns but not the BR ns (they are the same).

Reply:
Can't you rely on CSI's `podInfoForMount` feature (https://kubernetes-csi.github.io/docs/pod-info.html) to retrieve this value?